File: DB/Backend/database/models/monument.py
"""
Pydantic models for Monument data.
Mirrors Frontend/src/types/monument.ts — MonumentData and nested types.
"""

from typing import Optional, List, Tuple
from pydantic import BaseModel


class HotspotAnnotationResponse(BaseModel):
    """Matches Frontend HotspotAnnotation interface."""
    id: str
    title: str = ""
    shortDescription: str = ""
    detailedText: str = ""
    position: List[float] = []          # [x, y, z]
    cameraPosition: Optional[List[float]] = None
    cameraTarget: Optional[List[float]] = None
    imageUrl: Optional[str] = None
    architecturalNote: Optional[str] = None


class Monument3DModelConfigResponse(BaseModel):
    """Matches Frontend Monument3DModelConfig interface."""
    modelType: Optional[str] = None
    defaultCameraPosition: Optional[List[float]] = None
    lookAtTarget: Optional[List[float]] = None
    lightingPreset: Optional[str] = None
    minDistance: Optional[float] = None
    maxDistance: Optional[float] = None
    autoRotateSpeed: Optional[float] = None
    modelScale: Optional[float] = None
    groundOffset: Optional[float] = None


class MonumentResponse(BaseModel):
    """Full MonumentData — matches Frontend MonumentData interface."""
    id: str
    slug: str
    name: str
    nativeName: Optional[str] = None
    marathiName: Optional[str] = None
    hindiName: Optional[str] = None
    state: str = ""
    stateId: str = ""
    region: str = ""
    district_or_city: str = ""
    category: str = ""
    description: str = ""
    shortDescription: str = ""
    image: str = ""
    bannerImage: Optional[str] = None
    modelUrl: Optional[str] = None
    modelAvailable: bool = False
    architecturalStyle: Optional[str] = None
    yearBuilt: Optional[str] = None
    culturalSignificance: Optional[str] = None
    detailedHistory: Optional[str] = None
    culturalImportance: Optional[str] = None
    locationName: Optional[str] = None
    coordinates: Optional[dict] = None
    modelConfig: Optional[Monument3DModelConfigResponse] = None
    hotspots: Optional[List[HotspotAnnotationResponse]] = None
    starSchemaNodes: Optional[List[dict]] = None
    sources: Optional[List[dict]] = None


class MonumentSummaryResponse(BaseModel):
    """Lightweight summary for list endpoints."""
    id: str
    slug: str
    name: str
    stateId: str = ""
    state: str = ""
    shortDescription: str = ""
    image: str = ""
    modelAvailable: bool = False

File: DB/Frontend/src/types/monument.ts
import { SourceCitation, StarSchemaNode } from './culturalItem';

export interface HotspotAnnotation {
  id: string;
  title: string;
  shortDescription: string;
  detailedText: string;
  position: [number, number, number];
  cameraPosition?: [number, number, number];
  cameraTarget?: [number, number, number];
  imageUrl?: string;
  architecturalNote?: string;
}

export interface Monument3DModelConfig {
  modelType?: 'gltf' | 'procedural-arch' | 'custom-mesh';
  defaultCameraPosition?: [number, number, number];
  lookAtTarget?: [number, number, number];
  lightingPreset?: 'warm-heritage' | 'golden-hour' | 'museum-daylight' | 'dusk';
  minDistance?: number;
  maxDistance?: number;
  autoRotateSpeed?: number;
  modelScale?: number;
  groundOffset?: number;
}

export interface MonumentData {
  id: string;
  slug: string;
  name: string;
  nativeName?: string;
  marathiName?: string;
  hindiName?: string;
  state: string;
  stateId: string;
  region: string;
  district_or_city: string;
  category: string;
  description: string;
  shortDescription: string;
  image: string;
  bannerImage?: string;

  // Backend 3D Integration Contract
  modelUrl: string | null;
  modelAvailable: boolean;

  architecturalStyle?: string;
  yearBuilt?: string;
  culturalSignificance?: string;
  detailedHistory?: string;
  culturalImportance?: string;
  locationName?: string;
  coordinates?: {
    lat: number;
    lng: number;
  };
  modelConfig?: Monument3DModelConfig;
  hotspots?: HotspotAnnotation[];
  starSchemaNodes?: StarSchemaNode[];
  sources?: SourceCitation[];
}
